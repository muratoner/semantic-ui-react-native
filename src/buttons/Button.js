import PropTypes from 'prop-types';
import React from 'react';
import { StyleSheet, Text, TouchableOpacity } from 'react-native';
import Icon from '../icons/Icon';
import ButtonColor from './ButtonColor';

const Button = ({
	circular,
	outline,
	disabled,
	iconName,
	iconType,
	color,
	title,
	style,
	fluid,
	onPress
}) => {
	const bgColor = ButtonColor.getBgColor(color);
	const textColor = outline ? bgColor : ButtonColor.getTextColor(color);
	const iconStyle = {
		marginRight: title ? 5 : 0
	};
	return (
		<TouchableOpacity
			onPress={onPress}
			disabled={disabled}
			style={StyleSheet.flatten([
				style,
				{
					backgroundColor: outline ? '#fff' : bgColor,
					paddingHorizontal: 11,
					paddingVertical: 11,
					borderRadius: circular ? 50 : 3,
					alignItems: 'center',
					flexDirection: 'row',
					borderWidth: outline ? 1 : 0,
					borderColor: outline ? bgColor : null,
					opacity: disabled ? 0.45 : 1,
					flex: fluid ? 1 : null
				}
			])}
		>
			{iconName && iconType && (
				<Icon
					name={iconName}
					type={iconType}
					color={textColor}
					style={StyleSheet.flatten([styles.icon, iconStyle])}
				/>
			)}
			{title && (
				<Text
					style={{
						fontWeight: 'bold',
						color: textColor
					}}
				>
					{title}
				</Text>
			)}
		</TouchableOpacity>
	);
};

const styles = StyleSheet.create({
	icon: {
		fontSize: 15
	}
});

Button.propTypes = {
	title: PropTypes.string,
	color: PropTypes.oneOf([
		'primary',
		'secondary',
		'red',
		'positive',
		'negative'
	]),
	iconName: PropTypes.string,
	iconType: PropTypes.oneOf([
		'AntDesign',
		'Entypo',
		'EvilIcons',
		'Feather',
		'FontAwesome',
		'FontAwesome5',
		'Fontisto',
		'Foundation',
		'Ionicons',
		'MaterialCommunityIcons',
		'MaterialIcons',
		'Octicons',
		'SimpleLineIcons',
		'Zocial'
	])
};

export default Button;
