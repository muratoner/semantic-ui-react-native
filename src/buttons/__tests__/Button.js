import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import React from 'react';
import Button from '../Button';

describe('Button Component', () => {
	it('should render without issues', () => {
		const component = shallow(<Button />);

		expect(component.length).toBe(1);
		expect(toJson(component)).toMatchSnapshot();
	});

	it('should not be call onPress events when disabled is true', () => {
		const onPress = jest.fn();
		const component = shallow(<Button onPress={onPress} disabled />);

		expect(component.get(0).props.style).toHaveProperty('opacity', 0.45);

		// Simulate press action
		component.simulate('press');

		expect(onPress).not.toHaveBeenCalled();
	});

	it('should not be call onPress events when loading is true', () => {
		const onPress = jest.fn();
		const component = shallow(<Button onPress={onPress} loading />);

		expect(component.find('ActivityIndicator')).toHaveLength(1);

		// Simulate press action
		component.simulate('press');

		expect(onPress).not.toHaveBeenCalled();
	});

	it('should be call onPress events', () => {
		const onPress = jest.fn();
		const component = shallow(<Button onPress={onPress} />);

		// Simulate press action
		component.simulate('press');

		expect(onPress).toHaveBeenCalled();
	});

	it('should render with title prop', () => {
		const component = shallow(<Button title="Save" />);

		expect(component.length).toBe(1);
		expect(toJson(component)).toMatchSnapshot();
	});

	it('should render with title and circular props', () => {
		const component = shallow(<Button title="Save" circular />);

		expect(component.length).toBe(1);
		expect(toJson(component)).toMatchSnapshot();
	});

	it('should render with title, circular and disabled props', () => {
		const component = shallow(<Button title="Save" circular disabled />);

		expect(component.length).toBe(1);
		expect(toJson(component)).toMatchSnapshot();
	});

	it('should render with title, circular, disabled and fluid props', () => {
		const component = shallow(<Button title="Save" circular disabled fluid />);

		expect(component.length).toBe(1);
		expect(toJson(component)).toMatchSnapshot();
	});

	it('should render with title, circular, disabled, fluid and outline props', () => {
		const component = shallow(
			<Button title="Save" circular disabled fluid outline />
		);

		expect(component.length).toBe(1);
		expect(toJson(component)).toMatchSnapshot();
	});

	it('should render with title, circular, disabled, fluid, outline style props', () => {
		const component = shallow(
			<Button
				title="Save"
				circular
				disabled
				fluid
				outline
				style={{
					height: 50
				}}
			/>
		);

		expect(component.length).toBe(1);
		expect(toJson(component)).toMatchSnapshot();
	});

	it('should render with title, circular, disabled, fluid, outline and style props', () => {
		const component = shallow(
			<Button
				title="Save"
				circular
				disabled
				fluid
				outline
				style={{
					height: 50
				}}
			/>
		);

		expect(component.length).toBe(1);
		expect(toJson(component)).toMatchSnapshot();
	});

	it('should render with title, circular, disabled, fluid, outline, style and color props', () => {
		const component = shallow(
			<Button
				title="Save"
				circular
				disabled
				fluid
				outline
				style={{
					height: 50
				}}
				color="red"
			/>
		);

		expect(component.length).toBe(1);
		expect(toJson(component)).toMatchSnapshot();
	});

	it('should render with title, circular, disabled, fluid, outline, style and color props', () => {
		const component = shallow(
			<Button
				title="Save"
				circular
				disabled
				fluid
				outline
				style={{
					height: 50
				}}
				color="red"
				iconName="user"
			/>
		);

		expect(component.length).toBe(1);
		expect(toJson(component)).toMatchSnapshot();
	});

	it('should render with title, circular, disabled, fluid, outline, style, color and iconName props', () => {
		const component = shallow(
			<Button
				title="Save"
				circular
				disabled
				fluid
				outline
				style={{
					height: 50
				}}
				color="red"
				iconName="user"
			/>
		);

		expect(component.length).toBe(1);
		expect(toJson(component)).toMatchSnapshot();
	});

	it('should render with title, iconName and iconType props', () => {
		const component = shallow(
			<Button title="Add Friend" iconName="user" iconType="Ionicons" />
		);

		expect(component.length).toBe(1);
		expect(toJson(component)).toMatchSnapshot();
	});
});
