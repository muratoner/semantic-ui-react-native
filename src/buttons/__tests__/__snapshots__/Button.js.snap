// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button Component should render with title and circular props 1`] = `
<ForwardRef
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "#e0e1e2",
      "borderColor": null,
      "borderRadius": 50,
      "borderWidth": 0,
      "flex": null,
      "flexDirection": "row",
      "justifyContent": "center",
      "opacity": 1,
      "paddingHorizontal": 11,
      "paddingVertical": 11,
    }
  }
>
  <Text
    style={
      Object {
        "color": "rgba(0,0,0,.6)",
        "fontWeight": "bold",
      }
    }
  >
    Save
  </Text>
</ForwardRef>
`;

exports[`Button Component should render with title prop 1`] = `
<ForwardRef
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "#e0e1e2",
      "borderColor": null,
      "borderRadius": 3,
      "borderWidth": 0,
      "flex": null,
      "flexDirection": "row",
      "justifyContent": "center",
      "opacity": 1,
      "paddingHorizontal": 11,
      "paddingVertical": 11,
    }
  }
>
  <Text
    style={
      Object {
        "color": "rgba(0,0,0,.6)",
        "fontWeight": "bold",
      }
    }
  >
    Save
  </Text>
</ForwardRef>
`;

exports[`Button Component should render with title, circular and disabled props 1`] = `
<ForwardRef
  disabled={true}
  onPress={null}
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "#e0e1e2",
      "borderColor": null,
      "borderRadius": 50,
      "borderWidth": 0,
      "flex": null,
      "flexDirection": "row",
      "justifyContent": "center",
      "opacity": 0.45,
      "paddingHorizontal": 11,
      "paddingVertical": 11,
    }
  }
>
  <Text
    style={
      Object {
        "color": "rgba(0,0,0,.6)",
        "fontWeight": "bold",
      }
    }
  >
    Save
  </Text>
</ForwardRef>
`;

exports[`Button Component should render with title, circular, disabled and fluid props 1`] = `
<ForwardRef
  disabled={true}
  onPress={null}
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "#e0e1e2",
      "borderColor": null,
      "borderRadius": 50,
      "borderWidth": 0,
      "flex": 1,
      "flexDirection": "row",
      "justifyContent": "center",
      "opacity": 0.45,
      "paddingHorizontal": 11,
      "paddingVertical": 11,
    }
  }
>
  <Text
    style={
      Object {
        "color": "rgba(0,0,0,.6)",
        "fontWeight": "bold",
      }
    }
  >
    Save
  </Text>
</ForwardRef>
`;

exports[`Button Component should render with title, circular, disabled, fluid and outline props 1`] = `
<ForwardRef
  disabled={true}
  onPress={null}
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "#fff",
      "borderColor": "#e0e1e2",
      "borderRadius": 50,
      "borderWidth": 1,
      "flex": 1,
      "flexDirection": "row",
      "justifyContent": "center",
      "opacity": 0.45,
      "paddingHorizontal": 11,
      "paddingVertical": 11,
    }
  }
>
  <Text
    style={
      Object {
        "color": "#e0e1e2",
        "fontWeight": "bold",
      }
    }
  >
    Save
  </Text>
</ForwardRef>
`;

exports[`Button Component should render with title, circular, disabled, fluid, outline and style props 1`] = `
<ForwardRef
  disabled={true}
  onPress={null}
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "#fff",
      "borderColor": "#e0e1e2",
      "borderRadius": 50,
      "borderWidth": 1,
      "flex": 1,
      "flexDirection": "row",
      "height": 50,
      "justifyContent": "center",
      "opacity": 0.45,
      "paddingHorizontal": 11,
      "paddingVertical": 11,
    }
  }
>
  <Text
    style={
      Object {
        "color": "#e0e1e2",
        "fontWeight": "bold",
      }
    }
  >
    Save
  </Text>
</ForwardRef>
`;

exports[`Button Component should render with title, circular, disabled, fluid, outline style props 1`] = `
<ForwardRef
  disabled={true}
  onPress={null}
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "#fff",
      "borderColor": "#e0e1e2",
      "borderRadius": 50,
      "borderWidth": 1,
      "flex": 1,
      "flexDirection": "row",
      "height": 50,
      "justifyContent": "center",
      "opacity": 0.45,
      "paddingHorizontal": 11,
      "paddingVertical": 11,
    }
  }
>
  <Text
    style={
      Object {
        "color": "#e0e1e2",
        "fontWeight": "bold",
      }
    }
  >
    Save
  </Text>
</ForwardRef>
`;

exports[`Button Component should render with title, circular, disabled, fluid, outline, style and color props 1`] = `
<ForwardRef
  disabled={true}
  onPress={null}
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "#fff",
      "borderColor": "#db2828",
      "borderRadius": 50,
      "borderWidth": 1,
      "flex": 1,
      "flexDirection": "row",
      "height": 50,
      "justifyContent": "center",
      "opacity": 0.45,
      "paddingHorizontal": 11,
      "paddingVertical": 11,
    }
  }
>
  <Text
    style={
      Object {
        "color": "#db2828",
        "fontWeight": "bold",
      }
    }
  >
    Save
  </Text>
</ForwardRef>
`;

exports[`Button Component should render with title, circular, disabled, fluid, outline, style and color props 2`] = `
<ForwardRef
  disabled={true}
  onPress={null}
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "#fff",
      "borderColor": "#db2828",
      "borderRadius": 50,
      "borderWidth": 1,
      "flex": 1,
      "flexDirection": "row",
      "height": 50,
      "justifyContent": "center",
      "opacity": 0.45,
      "paddingHorizontal": 11,
      "paddingVertical": 11,
      "shadowColor": "#191919",
      "shadowOffset": Object {
        "height": 2,
        "width": 1,
      },
      "shadowOpacity": 0.2,
      "shadowRadius": 0.1,
    }
  }
>
  <Text
    style={
      Object {
        "color": "#db2828",
        "fontWeight": "bold",
      }
    }
  >
    Save
  </Text>
</ForwardRef>
`;

exports[`Button Component should render with title, circular, disabled, fluid, outline, style, color and iconName props 1`] = `
<ForwardRef
  disabled={true}
  onPress={null}
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "#fff",
      "borderColor": "#db2828",
      "borderRadius": 50,
      "borderWidth": 1,
      "flex": 1,
      "flexDirection": "row",
      "height": 50,
      "justifyContent": "center",
      "opacity": 0.45,
      "paddingHorizontal": 11,
      "paddingVertical": 11,
      "shadowColor": "#191919",
      "shadowOffset": Object {
        "height": 2,
        "width": 1,
      },
      "shadowOpacity": 0.2,
      "shadowRadius": 0.1,
    }
  }
>
  <Text
    style={
      Object {
        "color": "#db2828",
        "fontWeight": "bold",
      }
    }
  >
    Save
  </Text>
</ForwardRef>
`;

exports[`Button Component should render with title, iconName and iconType props 1`] = `
<ForwardRef
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "#e0e1e2",
      "borderColor": null,
      "borderRadius": 3,
      "borderWidth": 0,
      "flex": null,
      "flexDirection": "row",
      "justifyContent": "center",
      "opacity": 1,
      "paddingHorizontal": 11,
      "paddingVertical": 11,
    }
  }
>
  <Icon
    color="rgba(0,0,0,.6)"
    name="user"
    style={
      Object {
        "fontSize": 15,
        "marginRight": 5,
      }
    }
    type="Ionicons"
  />
  <Text
    style={
      Object {
        "color": "rgba(0,0,0,.6)",
        "fontWeight": "bold",
      }
    }
  >
    Add Friend
  </Text>
</ForwardRef>
`;

exports[`Button Component should render without issues 1`] = `
<ForwardRef
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "#e0e1e2",
      "borderColor": null,
      "borderRadius": 3,
      "borderWidth": 0,
      "flex": null,
      "flexDirection": "row",
      "justifyContent": "center",
      "opacity": 1,
      "paddingHorizontal": 11,
      "paddingVertical": 11,
    }
  }
/>
`;
